modules/generator/localentitylimiter/local_entity_limiter.go

```diff
@@ -0,0 +1,125 @@
+		activeEntities: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: "tempo",
+			Name:      "metrics_generator_registry_active_entities",
+			Help:      "The number of active entities in the metrics generator registry",
+		}, []string{"tenant"}),
+		maxActiveEntities: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: "tempo",
+			Name:      "metrics_generator_registry_max_active_entities",
+			Help:      "The maximum number of entities allowed to be active in the metrics generator registry",
+		}, []string{"tenant"}),
+		totalEntitiesAdded: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Namespace: "tempo",
+			Name:      "metrics_generator_registry_entities_added_total",
+			Help:      "The total amount of entities created per tenant",
+		}, []string{"tenant"}),
+		totalEntitiesRemoved: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Namespace: "tempo",
+			Name:      "metrics_generator_registry_entities_removed_total",
+			Help:      "The total amount of entities removed after they have become stale per tenant",
+		}, []string{"tenant"}),
+	}
+}
+
+var metrics = newMetrics(prometheus.DefaultRegisterer)
+
+type LocalEntityLimiter struct {
+	tenant            string
+	entityLastUpdated map[uint64]uint32
+	mtx               sync.Mutex
+	maxEntityFunc     func(tenant string) uint32
+	limitLogger       *tempo_log.RateLimitedLogger
+
+	metricTotalEntitiesLimited prometheus.Counter
+	metricActiveEntities       prometheus.Gauge
+	metricMaxActiveEntities    prometheus.Gauge
+	metricTotalEntitiesAdded   prometheus.Counter
+	metricTotalEntitiesRemoved prometheus.Counter
+}
+
+func New(maxEntityFunc func(tenant string) uint32, tenant string, limitLogger *tempo_log.RateLimitedLogger) *LocalEntityLimiter {
+	return &LocalEntityLimiter{
+		tenant:            tenant,
+		entityLastUpdated: make(map[uint64]uint32),
+		maxEntityFunc:     maxEntityFunc,
+		limitLogger:       limitLogger,
+
+		metricTotalEntitiesLimited: metrics.totalEntitiesLimited.WithLabelValues(tenant),
+		metricActiveEntities:       metrics.activeEntities.WithLabelValues(tenant),
+		metricMaxActiveEntities:    metrics.maxActiveEntities.WithLabelValues(tenant),
+		metricTotalEntitiesAdded:   metrics.totalEntitiesAdded.WithLabelValues(tenant),
+		metricTotalEntitiesRemoved: metrics.totalEntitiesRemoved.WithLabelValues(tenant),
+	}
+}
+
+func (l *LocalEntityLimiter) OnAdd(labelHash uint64, seriesCount uint32) bool {
+	l.mtx.Lock()
+	defer l.mtx.Unlock()
+
+	activeSeries, ok := l.entityLastUpdated[labelHash]
+	if ok {
+		activeSeries += seriesCount
+		l.entityLastUpdated[labelHash] = activeSeries
+		return true
+	}
+
+	maxEntities := l.maxEntityFunc(l.tenant)
+	if maxEntities != 0 && uint32(len(l.entityLastUpdated))+1 > maxEntities {
+		l.limitLogger.Log("msg", "reached max active entities", "active_entities", len(l.entityLastUpdated), "max_active_entities", maxEntities)
+		l.metricTotalEntitiesLimited.Add(float64(1))
```

Comment:
we could track series limited as well? i don't know if that's important to the way you want to operate this. up to you!